crates/bevy_render/src/color/mod.rs

```diff
@@ -1332,38 +1322,45 @@ impl encase::private::CreateFrom for Color {
 
 impl encase::ShaderSize for Color {}
 
-#[derive(Debug, Error)]
+#[derive(Debug, Error, PartialEq, Eq)]
 pub enum HexColorError {
     #[error("Unexpected length of hex string")]
     Length,
-    #[error("Error parsing hex value")]
-    Hex(#[from] hex::FromHexError),
+    #[error("Invalid hex char")]
+    Char(char),
 }
 
-fn decode_rgb(data: &[u8]) -> Result<Color, HexColorError> {
-    let mut buf = [0; 3];
-    match hex::decode_to_slice(data, &mut buf) {
-        Ok(_) => {
-            let r = buf[0] as f32 / 255.0;
-            let g = buf[1] as f32 / 255.0;
-            let b = buf[2] as f32 / 255.0;
-            Ok(Color::rgb(r, g, b))
-        }
-        Err(err) => Err(HexColorError::Hex(err)),
+// Convert hex to rgb[a] u8
+// For rgb: `fff` -> [u8; 6] -> [r, g, b, ..]
+// For rgba: `E2E2E2FF` -> [u8; 8] -> [r, g, b, a, ..]
+const fn decode_hex<const N: usize>(mut bytes: [u8; N]) -> Result<[u8; N], HexColorError> {
+    let mut i = 0;
+    while i < bytes.len() {
+        // Convert hex to u8
+        // e.g `f` -> 102 -> 15
+        let val = match hex_ascii_byte(bytes[i]) {
+            Ok(val) => val,
+            Err(byte) => return Err(HexColorError::Char(byte as char)),
+        };
+        bytes[i] = val;
+        i += 1;
     }
+    i = 0;
+    while i < bytes.len() / 2 {
+        // Convert u8 to r/g/b/a
+        // e.g `ff` -> [102, 102] -> [15, 15] = 255
+        bytes[i] = bytes[i * 2] << 4 | bytes[i * 2 + 1];
```

Comment:
Just write `bytes[i * 2] * 16`. It's trivial for a compiler to optimize. I don't know about `bytes[i * 2] * 16 + bytes[i * 2 + 1]`, though.